lnpbp-0021.md

```diff
@@ -56,102 +56,111 @@ Interface specification is the following Contractum code:
-    reserves TxOut,
-    proof [Byte] -- schema-specific proof 
-
-data Token ::
+data TokenData ::
     id TokenId,
-    name [Ascii ^ 1..40],
-    details [Unicode ^ 40..256]?,
-    preview Media?, -- always embedded preview media < 64kb
-    media Attachment?, -- external media which is the main media for the token
+    ticker RGBTypes.Ticker?,
+    name RGBTypes.Name?,
+    details RGBTypes.Details?,
+    -- always-embedded preview media < 64kb
+    preview EmbeddedMedia?,
+    -- external media which is the main media for the token
+    media Attachment?,
     attachments { U8 -> ^ ..20 Attachment } -- auxiliary attachments by type (up to 20 attachments)
-    reserves POR? -- output containing locked bitcoins; how reserves are
-                  -- proved is a matter of a specific schema implementation
-
-data Denomination :: 
-    ticker [Ascii ^ 1..8],
-    name [Ascii ^ 1..40],
-    details [Unicode ^ 40..256]?,
-    contract [Unicode]??,
+    -- output containing locked bitcoins; how reserves are proved is a matter
+    -- of a specific schema implementation
+    reserves RGBTypes.PoR? 
 
  -- each attachment type is a mapping from attachment id 
  -- (used as `Token.attachments` keys) to a short Ascii string
  -- verbally explaining the type of the attachment for the UI
  -- (like "sample" etc).
-data AttachmentType :: id U8, description [Ascii ^ 1..20]
-
+data AttachmentType :: 
+    id U8, 
+    description [Ascii ^ 1..20]
 
 interface RGB21
-    global Name :: Denomination
-    global Tokens :: Token+
-    global Engravings :: Engraving+
+    global name :: RGBTypes.AssetNaming
+    
+    -- Contract text is separated from the name since it must not be
+    -- changeable by the issuer.
+    global terms :: RGBTypes.RicardianContract
+
+    -- Data for all issued tokens
+    global tokensData+ :: TokenData
+    global engravingData+ :: EngravingData
     global isFractional :: Bool
-    global AttachmentTypes :: AttachmentType+
-
-    owned Allocations+ :: Allocation
-    owned IssueRight* :: Amount
-    owned DenominationRight?
-    -- returns information about known circulating supply
-    read supplyKnown :: Amount
-        count Self.state["Tokens"]
-        -- sum Self.ops["issue"]..closed["usingRight"].state ?? 0
-
-    -- maximum possible asset inflation
-    read supplyLimit :: Amount
-        sum Self.IssueRight..state ?? 0
-
-    read isCirculationKnown :: Bool
-        all Self.ops["issue"]..stateKnown
-
-    op transfer    :: inputs [Allocation+] -> beneficiaries [Allocation]
+    global attachmentTypes+ :: AttachmentType
+
+    -- Right to do a secondary (post-genesis) issue
+    owned inflationAllowance* :: RGBTypes.Amount
+    -- Right to update asset name
+    owned updateRight?
+
+    -- Ownership right over assets
+    owned assetOwner+ :: Allocation
+
+    genesis       -> name
+                   , terms,
+                   , reserves PoR*
+                   , tokens tokenData+
+                   , assetOwner+
```

Comment:
Shouldn't these be defined as a pair? IMO they should always match 1:1

Reply:
Nope, since it is impossible to mix global state with owned state into a single type (pair). The validation of 1:1 matching must be performed by a schema validation script

Reply:
Actually they are not a 1:1 pairs: tokens are the token data, and some of them may be fractionally-owned by multiple owners